src/modules/job-manager/update.c

```diff
@@ -588,11 +592,132 @@ static void send_error_responses (struct update *update)
     }
 }
 
+static void update_expiration_from_lookup_response (struct update *update,
+                                                    flux_future_t *f)
+{
+    flux_t *h = update->ctx->h;
+    const char *R;
+    json_t *o = NULL;
+    json_error_t error;
+
+    error.text[0] = '\0';
+
+    if (flux_kvs_lookup_get (f, &R) < 0
+        || !(o = json_loads (R, 0, NULL))
+        || json_unpack_ex (o, &error, 0,
+                           "{s:{s:F}}",
+                           "execution",
+                            "expiration", &update->instance_expiration) < 0)
+        flux_log (h,
+                  LOG_ERR,
+                  "failed to unpack current instance expiration: %s",
+                  error.text);
+    json_decref (o);
+}
+
+/*  An update to resource.R has occurred. If expiration > previous
+ *  expiration, then adjust expiration of all running jobs where
+ *  no duration is set in jobspec, but the job currently has a
+ *  set expiration. This implies the expiration was set automatically
+ *  by the scheduler and needs an update.
+ */
+static void resource_update_cb (flux_future_t *f, void *arg)
+{
+    struct update *update = arg;
+    flux_t *h = update->ctx->h;
+    struct job *job;
+    zlistx_t *jobs;
+    double old_expiration = update->instance_expiration;
+
+    update_expiration_from_lookup_response (update, f);
+    flux_future_reset (f);
+
+    /*  If this is the first successful update, or the expiration was
+     *  reduced, or there are no running jobs, then there is nothing
+     *  more to do.
+     */
+    if (old_expiration == -1.
+        || (update->instance_expiration > 0.
+            && update->instance_expiration < old_expiration)
+        || update->ctx->running_jobs == 0)
+        return;
```

Comment:
I think you said before that the reasoning for not reducing the expiration time on jobs when the enclosing instance expiration is reduced was that there would be no material difference in behavior since running jobs are canceled as part of instance shutdown anyway.

Would a comment about that here be useful to future explorers wondering why we don't do that?

Another useful comment might be that the motivating use case is "hotline extending the duration of a running batch job" or similar.

Reply:
Actually, now that it's been said out in the open, the current solution doesn't sound right. For one thing, `flux job timeleft` will report the incorrect value, and if the job itself is an instance, it should probably have its resource set expiration reduced so that new submitted jobs don't inherit the incorrect expiration. Not sure why I thought this was a good idea at first.

I'll remove that special case, add the suggested comment, and also add a test for reducing the expiration of a instance...